### ISSUE TITLE
aws_instance.volume_tags gets confused when > 1 volume attached

### ISSUE THREAD

**Issue by hashibot** [hashicorp/terraform-provider-aws#729]
_This issue was originally opened by @FransUrbo as hashicorp/terraform#14107. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

It seems the new `volume_tags` (thanx btw!) gets confused when there's more than one volume attached to an instance.

### Terraform Version
0.9.4

### Affected Resource(s)
Please list the resources as a list, for example:
- aws_instance

### Terraform Configuration Files
```hcl
resource "aws_instance" "instance" {
  root_block_device {
    volume_type               = "standard"
    volume_size               = 8
  }

  volume_tags {
    Name                      = "my-instance-root"
  }
}

resource "aws_ebs_volume" "instance" {
  tags {
    Name                      = "my-instance-db"
  }
}

resource "aws_volume_attachment" "instance" {
  instance_id                 = "${aws_instance.instance.id}"
  volume_id                   = "${aws_ebs_volume.instance.id}"
  device_name                 = "/dev/sdb"
}
```

### Expected Behavior
Nothing.

### Actual Behavior
Every other time it modified the `aws_ebs_volume` tags and every other on the `aws_instance`.

### Steps to Reproduce
1. `terraform apply`

**Comment by kpumuk** [comment]
👍  have the same issue

**Comment by rickhlx** [comment]
+1

**Comment by acutchin-bitpusher** [comment]
still seeing this in 0.10.8

**Comment by e-carlin** [comment]
+1

**Comment by MrMojoRisin49** [comment]
+1

**Comment by ddsdevon** [comment]
Still an issue as of terraform 0.11.1
I added a second volume to an existing instance with its own tags, and now aws_instance and aws_ebs_volume take turns performing an update in place to toggle the tags back and forth.

**Comment by geekbass** [comment]
Upgraded to version 0.11.1 and I am seeing this issue as well.

**Comment by jdbeyers** [comment]
+1

**Comment by SYC1205** [comment]
+1

version 0.11.1 still has this issue

**Comment by chan-alex** [comment]
+1 
Seeing this in version 0.11.1

**Comment by smastrorocco** [comment]
+1
Seeing this in version 0.11.2

**Comment by duhaas2015** [comment]
I can confirm I'm seeing the same behavior in 0.11.3, adding the ignore for volumes tags to the lifecycle fixes the problem for now

**Comment by gavD** [comment]
I am still getting this problem

```
$ terraform --version
Terraform v0.11.7
+ provider.aws v1.7.0
```

My code (simplified) looks like:

```
resource "aws_instance" "allinone" {
  ami                     = "${var.ami}"
  subnet_id               = "${var.subnet_id}"

  root_block_device {
    volume_type = "gp2"
  }

  volume_tags {
    MakeSnapshot   = "False"
  }
}

resource "aws_ebs_volume" "myebs" {

  type              = "gp2"
  size              = "${var.db_data_volume_size}"
  encrypted         = true

  tags {
    MakeSnapshot   = "True"
  }

}
```

I find that the tag `MakeSnapshot` on `myebs` vacillates between `False` and `True` with every `terraform apply`. I've tried making the EBS `depends_on` the EC2 instance and vica versa and it doesn't help.

Hope that is a clear explanation - it would be great to get a fix on this! :)

**Comment by roberulloa** [comment]
`$ terraform --version
Terraform v0.11.7
+ provider.aws v1.18.0`

This issue comes up when you use the volume_tags inside aws_instance with the aws_ebs_volume tag option.
` volume_tags {
    Name = "ebs-root-instance"
  }`

If you use only the tag mode on an aws_ebs_volume, works like a charm.

**Comment by asandusbux** [comment]
any solution with using _volume_tags_ inside aws_instance, from not overriding ebs volume tags? No issues with EBS volume tagging. I want to give a tag to the instance root volume using _volume_tags_

> terraform version: 
> Terraform v0.11.7

**Comment by mda590** [comment]
+1!

Seeing this exact behavior in version:
Terraform v0.11.7
+ provider.aws v1.23.0

**Comment by tdmalone** [comment]
Related tickets: #770 & #884 (possibly all duplicates of the same issue?)

**Comment by justin-sunayu** [comment]
+1 Getting this while trying to tag my root volume and data volumes

**Comment by gnalawade** [comment]
same issue observed
[gnalawade@jhost db]$ terraform -v
Terraform v0.11.7
+ provider.aws v1.26.0


aws_volume_attachment.sql101-data01-attach: Refreshing state... (ID: vai-1256441354)

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ aws_instance.sql101
      volume_tags.Name: "SQL101.PROD-sdf" => "SQL101.PRDO-sda"

Plan: 0 to add, 1 to change, 0 to destroy.

**Comment by spommerening** [comment]
Same issue here with
```
Terraform v0.11.7
+ provider.aws v1.29.0
```

**Comment by tdmalone** [comment]
Hi @soar @gnalawade @justin-sunayu @n3ph @mda590 

The issue will still exist until it’s fixed. There’s no need to add additional comments confirming it’s still present - if you’re affected, add a thumbs up reaction to the initial post in this thread instead.

Comments that don’t include additional information such as troubleshooting, workarounds etc add noise for the subscribers and risk being locked by the maintainers. It also makes it more difficult to find discussion of the actual issue.

Thanks!

**Comment by vfoucault** [comment]
+1

**Comment by morokin** [comment]
In the meanwhile, year 2019,
terraform --version
Terraform v0.11.13
+ provider.aws v2.2.0

Issue still there

**Comment by tdmalone** [comment]
@morokin Add a plus one on the initial post in this thread instead - Hashicorp report on those to determine which features to work on. As I mentioned in my above comment that you -1'ed, your post just adds needless noise for others in the same position who are subscribed to this thread, and doesn't actually help the issue get fixed or prioritised.

**Comment by bkmeneguello** [comment]
My suggestion is to add "tags" attribute to both "root_ebs_volume" and "ebs_block_device" while keeping the "volume_tags". This way "volume_tags" could be a fallback when no tags are defined in the previous elements (and to retain compatibility). If someone don't want all instance block devices tagged with "volume_tags" just suppress it and tag blocks individually.

**Comment by philiphope** [comment]
2 years. How long do we have to wait for this bug to be fixed?

**Comment by rifelpet** [comment]
Just for reference to anyone wishing to fix this, the problem stems from the fact that when [updating volume_tags in place](https://github.com/terraform-providers/terraform-provider-aws/blob/8108c11b07f7b6b53c93dab9ab501040bbb82aa0/aws/resource_aws_instance.go#L907-L912), terraform sets tags on [all volume IDs attached to the instance](https://github.com/terraform-providers/terraform-provider-aws/blob/a66865cc29b5b1bc2543b61b379368e3bd82ba60/aws/tags.go#L88-L131) rather than only the volumes created from the original [ec2.RunInstances](https://github.com/terraform-providers/terraform-provider-aws/blob/8108c11b07f7b6b53c93dab9ab501040bbb82aa0/aws/resource_aws_instance.go#L521) used to create the instance. Terraform is already able to [distinguish the root_block_device](https://github.com/terraform-providers/terraform-provider-aws/blob/8108c11b07f7b6b53c93dab9ab501040bbb82aa0/aws/resource_aws_instance.go#L1332-L1333) from other block devices but not distinguishing ebs_block_devices defined in an aws_instance versus a separate aws_ebs_volume resource. The ec2.RunInstances call does return the volume IDs of the other block devices, but any fix to track this information and use it would only fix it for instances created with a version of the aws provider containing the fix, not existing instances nor imported instances.

You can confirm this ambiguity with a `terraform show`. This `aws_instance` has a `root_block_device` but no `ebs_block_device`, rather it has separate `aws_ebs_volume` and `aws_volume_attachment` resources. The separate volume resource's volume ID is now stored as an attribute of the aws_instance:

```
aws_instance.foo:
  id = i-0e2ee1ae34f051b98
  ebs_block_device.# = 1
  ebs_block_device.4138763540.delete_on_termination = false
  ebs_block_device.4138763540.device_name = /dev/xvdq
  ebs_block_device.4138763540.volume_id = vol-07b923a8c02cc95f0
  ebs_block_device.4138763540.volume_size = 750
  ebs_block_device.4138763540.volume_type = gp2
  ebs_optimized = true
  ...
  root_block_device.# = 1
  root_block_device.0.delete_on_termination = true
  root_block_device.0.iops = 0
  root_block_device.0.volume_id = vol-0aed163a5aa8bf9d8
  root_block_device.0.volume_size = 20
  root_block_device.0.volume_type = standard
aws_ebs_volume.foo:
  id = vol-07b923a8c02cc95f0
  ...
aws_volume_attachment.foo:
  id = vai-1985794965
  device_name = /dev/xvdq
  instance_id = i-0e2ee1ae34f051b98
  skip_destroy = true
  volume_id = vol-07b923a8c02cc95f0
```

It would be pretty straightforward to use the [readBlockDevicesFromInstance](https://github.com/terraform-providers/terraform-provider-aws/blob/8108c11b07f7b6b53c93dab9ab501040bbb82aa0/aws/resource_aws_instance.go#L1283) function to get the root_block_device's volume ID for [setVolumeTags](https://github.com/terraform-providers/terraform-provider-aws/blob/a66865cc29b5b1bc2543b61b379368e3bd82ba60/aws/tags.go#L81) but that wouldnt solve the ambiguity with `ebs_block_device`. Personally, I'm a fan of this partial fix.

Thoughts?

**Comment by slessardjr** [comment]
I just overwrote the name of all my kubernetes persistent EBS volumes due to this bug.  Is there any plan to fix this?  I think the suggestion @rifelpet is a perfect middle ground. Unless the provider makes root_volume_tags and ebs_volume_tags separated.

**Comment by muhmud** [comment]
> 2 years. How long do we have to wait for this bug to be fixed?

+ another year. 

I just started using Terraform, and it's fantastic, but can we not get some traction on this?

**Comment by bkmeneguello** [comment]
I think the proper way is to submit a PR, it should not be that hard.

**Comment by binlab** [comment]
```shell
$ terraform version
Terraform v0.12.24
+ provider.aws v2.53.0
```
the problem still remains

**Comment by alanbantuit** [comment]
$ terraform --version
Terraform v0.12.26
+ provider.aws v2.66.0

still wants to change tags on the non-root volume:

      ~ volume_tags                  = {
          ~ "Backup" = "dev_de_data" -> "dev_de_root"
          ~ "Name"   = "DEV|DE|Data" -> "DEV|DE|Root"
        }

That apparently just 1 line changed pull request has still not been merged.

**Comment by thomasbiddle** [comment]
Running into this now, still.

```
Terraform v0.13.3
+ provider registry.terraform.io/-/aws v3.8.0
+ provider registry.terraform.io/hashicorp/aws v3.8.0
```

Having an option to just ignore all EBS volumes would be the best solution IMO.

**Comment by jtsoi** [comment]
I think https://github.com/hashicorp/terraform-provider-aws/pull/15474 is a good way to fix this.
Until then, here is a work around:


```
resource "aws_instance" "this" {
 ...
  # Volume
  root_block_device {
    volume_type           = "gp3"
    volume_size           = var.volume_size
    encrypted             = true
    delete_on_termination = true
  }
  # Instance tags
  tags = { ... }

  # Dont define 'volume_tags' here.
  ...
}

data aws_ebs_volume boot_volume {
  filter {
    name   = "attachment.device"
    values = ["/dev/sda1"]
  }
  filter {
    name   = "attachment.instance-id"
    values = [aws_instance.this.id]
  }
}

resource "aws_ec2_tag" boot_volume_tags {
  for_each  = var.boot_volume_tags
  resource_id = data.aws_ebs_volume.boot_volume.volume_id
  key         = each.key
  value       = each.value
}
```

Then attach additional volumes with 'aws_volume_attachment' and tags on those volumes will not change.

**Comment by bevanbennett** [comment]
Just as a note, this is now an issue as of 0.14 (maybe 0.13, we came direct from 0.12) for people who have never set volume_tags.
All my plans are now alternating between the aws_instance volume_tags deleting tags off my extra volumes and the tags specified in aws_ebs_volume putting them back. I'm going to try that workaround, but it's a LOT of new annoyance to implement over our hundreds of modules.

**Comment by bevanbennett** [comment]
Confirmed the workaround no longer works in 0.14.4.
Now plans alternate between aws_ec2_tagadding the tags and BOTH aws_ebs_volume (with no tags block) and aws_instance (with no volume_tags block) removing them.

**Comment by bevanbennett** [comment]
I can get the old behavior by specifying tags in the aws_ebs_volume resource and putting ignore_changes = [ volume_tags ] into the aws_instance, but it feels like a hack.

**Comment by alanbantuit** [comment]
That hack is what I have been doing while waiting for https://github.com/terraform-providers/terraform-provider-aws/pull/12226. It's been almost a year for a 1 line code fix.

________________________________

**Comment by YakDriver** [comment]
We have merged a fix to the `volume_tags` issue in #15474. We have added tests to cover the issues observed. Please note that using `volume_tags` in `aws_instance` is _not_ compatible with using `tags` in `aws_ebs_volume`. You need to use one or the other. Prior to this fix, even following this rule, you would encounter errors. Along with the fix, we've added `tags` to the `root_block_device` and `ebs_block_device` configuration blocks in `aws_instance`.

Now that the fix is in place, if you find any problems with `volume_tags`, let us know by [opening a new issue](https://github.com/hashicorp/terraform-provider-aws/issues/new/choose).

**Comment by brettryan** [comment]
> @morokin Add a plus one on the initial post in this thread instead - Hashicorp report on those to determine which features to work on. As I mentioned in my above comment that you -1'ed, your post just adds needless noise for others in the same position who are subscribed to this thread, and doesn't actually help the issue get fixed or prioritised.

This is true, however; if I'm not mistaken issues tend to get closed if you don't comment.
